### ISSUE TITLE
Implement kernel profiling

### ISSUE THREAD

**Issue by fknorr** [AdaptiveCpp/AdaptiveCpp#417]
hipSYCL is currently missing an API for kernel profiling. I would like to implement this, but I think it warrants some discussion beforehand (and I also need some directions for navigating the runtime internals).

For reference, there was an attempt at implementing this for the old runtime #251, but that PR has since become obsolete.

## Problems with profiling in the SYCL spec

The spec mandates `event::get_profiling_info`, which is basically the same interface as [`clEventGetProfilingInfo`](https://www.khronos.org/registry/OpenCL/sdk/1.0/docs/man/xhtml/clGetEventProfilingInfo.html). This makes it easy to implement for an OpenCL backend, where a SYCL event corresponds to an OpenCL event exactly and OpenCL events already report submission, start and end times.

CUDA profiling works differently. To measure the device execution time of a command, one needs to call `cudaEventRecord` before and after the kernel launch, and obtain the duration via `cudaEventElapsedTime(endEvt, startEvt)`. hipSYCL already records a CUDA event for the kernel completion, and when queue profiling is enabled, will have to insert an additional event before kernel launch.

But even with multiple CUDA events, we cannot implement the spec API faithfully. We don't have access to timestamps, only durations, so we'd have to report `command_start` as constant-zero and `command_end` as the duration. That might be enough for simple kernel profiling, but losing the temporal ordering between different command groups might surprise users. Also, I see no way to measure `command_submit` in device time at all.

The spec is also not too precise on when those individual timings should be taken. It talks about profiling a command group, which can be understood as including times for implicit data transfers, but the OpenCL-like interface suggestst that only kernel times are measured.

## To be discussed

1. Should we add our own extension for kernel profiling that can actually be implemented correctly with CUDA? I'm thinking along the lines of an `event::get_execution_time()` or `event::get_profiling_info<info::execution_time>`.
2. If we do add an extension, should we partially implement the spec API subject to the limitations above? This would improve conformance, but a partially-correct implementation might be worse than none at all for users migrating from a different SYCL implementation.
3. Should we only record kernel execution (or explicit-copy) times, or should requirement / implicit transfer durations be included in the profile? Regardless of how we decide, the extension API (if we want to implment it) should be unambiguous in that regard.

## Directions for an implementation

Where should these profiling events be created? There are code comments mentioning that `backend_executor::create_event_{before,after}` is meant for profiling in the scheduler, but I'm not sure where to apply that. I think we should only add events for nodes that do return a `sycl::event`, there is no need for profiling functionality e.g. on implicit transfer operations.

**Comment by illuhad** [comment]
Thanks for opening this discussion!

> Should we add our own extension for kernel profiling that can actually be implemented correctly with CUDA? I'm thinking along the lines of an event::get_execution_time() or event::get_profiling_info<info::execution_time>.

I see measuring of execution times as part of a larger instrumentation API that would, among other things, support measuring kernels runtimes, but could additionally expose additional information (e.g. using backend-specific frameworks such as cuPTI). As such, I think formulating this inside the hipSYCL runtime as duration (which we can measure well) and a start time (which some backends may not be able to expose in a precise fashion) makes sense.
I'm not opposed to exposing the duration as an extension in `sycl::event`, but since this affects CUDA backends in other implementations as well, it might be a candidate to discuss inside the SYCL wg if this needs to be changed.

> If we do add an extension, should we partially implement the spec API subject to the limitations above? This would improve conformance, but a partially-correct implementation might be worse than none at all for users migrating from a different SYCL implementation.

Yes. Actually, I don't see much of an issue here. We could just reuse the submission time for `command_start` as well. The concept of a device is loosely defined and specific to SYCL implementations and backends, so we could just as well interpret this as "when we hand off control", i.e. hand execution over to CUDA.

>  Should we only record kernel execution (or explicit-copy) times, or should requirement / implicit transfer durations be included in the profile? Regardless of how we decide, the extension API (if we want to implment it) should be unambiguous in that regard.
> Where should these profiling events be created?

Here's how I envision the instrumentation API and how profiling would fit in. The hipSYCL runtime has a `hints` mechanism (which is effectively similar to `sycl::property_list`) that can be used to pass instructions and hints from the SYCL interface to the runtime. The queue attaches a set of default hints (based on how it was constructed) to the operations submitted to it. These hints are then attached to DAG nodes, so the implementation of the instrumentation API needs to center on DAG nodes.

Requests for instrumentation (such as profiling) would be passed to the runtime by attaching a hint with the requested kind of instrumentation to the DAG nodes. If the queue is constructed with with the profiling property, it will attach the corresponding hints to all operations that go through it, including implicit data transfers (which, like all operations, have an event inside the hipSYCL runtime, even if it is maybe not exposed on the SYCL level). Not instrumenting data transfers would require extra handling of all these cases inside the `handler` to remove those hints again.

If it is not desired to instrument every operation, then there's another solution coming: I will very shortly publish the first hipSYCL extensions using an additional `property_list` parameter for individual command groups, i.e. the `submit()` API. We could also add a profiling property there, which would allow us to decide at the granularity of individual command groups whether we want to profile it.

In any case, since the `dag_node_event` inside a `sycl::event` only refers to a single operation, imo it only makes sense to measure the duration of a single operation and *not add* the duration of data transfers to those of kernels even if they were part of the same command group.

The DAG node then would go through the scheduler and arrive at the `backend_executor`. While we currently only have the `multi_queue_executor`, it is important to note that the CUDA/HIP queue and event model are *not* guaranteed in hipSYCL anymore as I would like to add additional task-graph based backends in the future that have different `backend_executor` implementations. So the profiling implementation should not assume these models blindly.

At the moment it seems most reasonable to me to let the backend handle instrumentation entirely, and, even in the `multi_queue_executor` case, not assume that the backend event i.e. `dag_node_event` is in any way related to profiling. For example, on CPU we could implement profiling very differently and directly measure timestamps at the beginning and end our operations without constructing events.
Because of this, I'm currently thinking in the direction of a mechanism where the backend is given the DAG node and then does whatever it needs to profile it. This could then require submitting additional events. At the moment, the `multi_queue_executor` does not hand the DAG node itself to the backend (only the operation description), so the design needs some thinking here.
Maybe we add an `instrumentation_state` class to the dag node, and we then bind this object to the `operation` so that the backend can see it during execution and take appropriate steps such as submitting additional events and providing a mechanism to access the results.


> There are code comments mentioning that backend_executor::create_event_{before,after} is meant for profiling in the scheduler, but I'm not sure where to apply that. 

I *think* these are only part of the `submit_dag()` interface that is unfinished/in flux and not currently in use and will probably be rewritten anyway. If this is the case we don't need to rely on them.

**Comment by fknorr** [comment]
Thanks for the detailed input!

> I see measuring of execution times as part of a larger instrumentation API that would, among other things, support measuring kernels runtimes, but could additionally expose additional information (e.g. using backend-specific frameworks such as cuPTI).

I really like that prospect. We should be able to implement the simple profiling requirements in the spec in a way that is extensible in the runtime later.

My current idea is to attach an opaque `profiler` struct to `dag_node` which will be instantiated by the backend `inorder_queue` if profiling is requested through an `execution_hint`. That profiler would contain a bunch of `cuEvent_t` for CUDA or a `std::chrono::time_point` for OpenMP. The profiler could be set either through an out-parameter at `operation::dispatch` / `inorder_queue::submit_*` or through a field in `operation`. In the latter case, we should probably make the operation parameter non-const on the backend queues (the operation is mutable further up the call chain anyway). Seeing an `operation` as a (logically-const) task description, the explicit out-parameter feels less tacked-on.

Now I'm not 100% sure at which point the profiler-pointer should be constructed (i.e. where I should put the `&node->profiler`, from which point I pass the profiler-pointer downwards). There is `dag_node::get_operation`, but `rt::submit` and `backend_executor::submit_directly` take separate `dag_node_ptr` and `operation` parameters, which  suggests that there can be more than one operation per `dag_node` or vice versa. I want to make sure that each node is profiled at most once.

For some input on the absolute-time issue I looked at [Intel's implementation for DPC++](https://github.com/intel/llvm/blob/sycl/sycl/plugins/cuda/pi_cuda.cpp). They submit additional events when the queue is created, which allows total ordering of time stamps at least within the same queue. That is enough to satisfy the spec, so we can actually stick to the SYCL API on that matter!

**Comment by illuhad** [comment]
> I really like that prospect. We should be able to implement the simple profiling requirements in the spec in a way that is extensible in the runtime later.

Agreed.

> My current idea is to attach an opaque profiler struct to dag_node which will be instantiated by the backend inorder_queue if profiling is requested through an execution_hint

Maybe we could have an abstract instrumentation class with virtual member functions for all the instrumentations that we support that can be implemented by backends?
Letting the `inorder_queue` handle this seems like the right approach to me. It's still unclear how the `inorder_queue` would get access to the hints to notice that instrumentation is requested since those are attached to the `dag_node`.

> . In the latter case, we should probably make the operation parameter non-const on the backend queues (the operation is mutable further up the call chain anyway). Seeing an operation as a (logically-const) task description, the explicit out-parameter feels less tacked-on.

Not sure, but I wonder if the operation might actually be the more natural way for instrumentation as it refers to an individual operation on the backend level which actually can be instrumtented, while a `dag_node` might be more abstract. So, if we mark the operation itself for instrumentation depending on the `dag_node` hints, it might also make sense to store the results there.

> Now I'm not 100% sure at which point the profiler-pointer should be constructed (i.e. where I should put the &node->profiler, from which point I pass the profiler-pointer downwards). There is dag_node::get_operation, but rt::submit and backend_executor::submit_directly take separate dag_node_ptr and operation parameters, which suggests that there can be more than one operation per dag_node or vice versa. I want to make sure that each node is profiled at most once.

For the operations that you most likely care about (e.g. kernels) there is only a single operation. Currently, if hipSYCL hits cases where it would need to have multiple operations, it generally errors because this is not yet fully supported. Multiple operations per dag node can only happen when translating implicit requirements into actual data transfers:
* when data is scattered across multiple devices, so multiple copies are necessary to update the data region an accessor is referring to
* when up-to-date data on a device is in non-rectangular shape, so multiple rectangular copies are necessary to migrate it
* when data has to be copied between two backends that cannot talk directly to each other (e.g. CUDA and ROCm), so copies have to go through the host with an additional data transfer. This is not really implemented since CUDA and ROCm currently cannot be simultaneously targeted anyway.

The idea is that if an operation is an implicit requirement, `submit_directly` expands it into one (or more) actual operation. These operations instead of the original one are then given together with the original DAG node to the backend executor.

The only case I can think of where the user actually gets an `event` for such a situation where he could query profiling information is `handler::update_host()`. Although it might be interesting to expand that API in the future and have a `handler::update_device(const device&)` as well.

I believe that ideally, the instrumentation class should be instantiated inside backend-specific code, i.e. the `inorder_queue` if possible, so that we can easily add backend-specific behavior.

> For some input on the absolute-time issue I looked at Intel's implementation for DPC++. They submit additional events when the queue is created, which allows total ordering of time stamps at least within the same queue. That is enough to satisfy the spec, so we can actually stick to the SYCL API on that matter!

Ingenious! It shouldn't be a problem to just have an event recorded in the backend queue constructor.

In summary, the open questions I see are:
* How could the backend `inorder_queue` know if instrumentation has been requested with a hint?
* Do we attach the instrumentation API to the dag node or the operation? How does it work when an implicit requirement operation has been translated into multiple actual operations (or do we even care about these cases?)?
   * If we attach to a DAG node, how can the backend `inorder_queue` access it (related to first question)?

**Comment by fknorr** [comment]
After experimenting with both approaches (explicit profiler-future parameter vs. embedding profiling functionality in `operation`), operation feels like the more natural place to attach profiling: Even if the queue (or, in the future, an individual submission) is profiled, its implicit requirements do not need to be profiled. The decision between explicit and implicit operations is easy when the operation is constructed, but difficult once the only interface is a `dag_node`.

On the topic of profiling vs. broader instrumentation, I believe that an abstract profiler interface that is extensible in the case when we decide to evaluate things like GPU performance counters is the best option. I'm hesitant to implement a larger, generic instrumentation API - without really knowing the requirements, that will probably lead to overengineering.

**Comment by illuhad** [comment]
>  Even if the queue (or, in the future, an individual submission) is profiled, its implicit requirements do not need to be profiled. 

Well... I'm not sure that statement is correct in general. E.g. `update_host()` uses this mechanism. Also, one might argue that requirements *are* part of a command group and therefore should be profiled. But I do agree that `operation` is probably still the right way to do this. If this is necessary in the future the solution would be to make the `dag_node` aware of all operations that have actually been submitted for it, which would then make them accessible.

> On the topic of profiling vs. broader instrumentation, I believe that an abstract profiler interface that is extensible in the case when we decide to evaluate things like GPU performance counters is the best option. I'm hesitant to implement a larger, generic instrumentation API - without really knowing the requirements, that will probably lead to overengineering.

I disagree because I believe it is straight-forward, e.g. something like:
```cpp
enum class instrumentation_type
{
  command_group_start_time,
  command_group_end_time,
  ...
};
// abstract API that can be implemented by backends
class instrumentation
{
public:
   virtual result enable(instrumentation_type t) = 0;
   virtual bool is_enabled(instrumentation_type t) const = 0;
   // assume instrumentation results can be expressed by double for now
   // if this isn't the case we can always have dedicated abstract member functions
   // for such an instrumentation
   virtual result get_result(instrumentation_type t, double& out) const = 0;
};

class operation {
  bool is_instrumented() const { return _instr; }
  // Set _instr
  void instrument(std::unique_ptr<intrumentation> instr);
private:
  std::unique_ptr<instrumentation> _instr;
};

```

**Comment by fknorr** [comment]
Ah, that doesn't look so bad, I was trying to wrap my head around an implementation that generalizes on profiling but uses an inheritance-based mechanism simlar to my proposal in #428.

The `instrumentation` structure you suggest reminds me of how the `execution_hint` mechanism is implemented: A fixed, but extensible set of instrumentations can be inserted / retrieved through template functions. This enables strong typing on the individual instrumentations and solves the problem of having to express every instrumentation result as `double`:

```c++
class device_timestamp_instrumentation {
public:
  enum class event { operation_submitted, operation_started, operation_finished };
  virtual profiler_clock::time_point await_event(event event) = 0;
};

template<typename Instr>
class future_instrumentation {
public:
  void provide(std::unique_ptr<Instr> instr);
  Instr &await();

private:
  std::mutex _mutex;
  std::condition_variable _update;
  std::unique_ptr<Instr> _resolved;
};

using any_future_instrumentation = std::variant<
    future_instrumentation<device_timestamp_instrumentation>
>;

class operation {
public:
  template<typename Instr> bool has_instrumentation() const;
  template<typename Instr> future_instrumentation<Instr> &get_instrumentation();
  template<typename Instr> void instrument();

private:
  std::vector<any_future_instrumentation> _instrs;
};
```

We could also implement this through `std::any` instead of `std::variant` to remove `operation`'s dependency on the set of all instrumentations. I'd still add some concept-like static assertion to stop the consumer from accidentally passing unrelated types as `Instr` in that case.

**Comment by FreLindberg** [comment]
I was interested in using the features implemented in #428 and was wondering if there is a specific reason as to why it hasn't been merged into the develop branch?

**Comment by illuhad** [comment]
@FreLindberg I think we hadn't finalized design discussions when we were interrupted by new year/christmas vacation, and afterwards it kind of got lost - time to pick it up again :)
